frame/assets/src/mock.rs

```diff
@@ -101,44 +101,51 @@ impl Config for Test {
 	type Extra = ();
 }
 
-use std::{cell::RefCell, collections::HashMap};
+use std::collections::HashMap;
 
 #[derive(Copy, Clone, Eq, PartialEq, Debug)]
-pub(crate) enum Hook {
+pub enum Hook {
 	Died(u32, u64),
 }
-thread_local! {
-	static FROZEN: RefCell<HashMap<(u32, u64), u64>> = RefCell::new(Default::default());
-	static HOOKS: RefCell<Vec<Hook>> = RefCell::new(Default::default());
+parameter_types! {
+	static Frozen: HashMap<(u32, u64), u64> = Default::default();
+	static Hooks: Vec<Hook> = Default::default();
 }
 
 pub struct TestFreezer;
 impl FrozenBalance<u32, u64, u64> for TestFreezer {
 	fn frozen_balance(asset: u32, who: &u64) -> Option<u64> {
-		FROZEN.with(|f| f.borrow().get(&(asset, *who)).cloned())
+		Frozen::get().get(&(asset, *who)).cloned()
 	}
 
 	fn died(asset: u32, who: &u64) {
-		HOOKS.with(|h| h.borrow_mut().push(Hook::Died(asset, *who)));
+		Hooks::mutate(|v| v.push(Hook::Died(asset, *who)));
+
 		// Sanity check: dead accounts have no balance.
 		assert!(Assets::balance(asset, *who).is_zero());
 	}
 }
 
 pub(crate) fn set_frozen_balance(asset: u32, who: u64, amount: u64) {
-	FROZEN.with(|f| f.borrow_mut().insert((asset, who), amount));
+	Frozen::mutate(|v| {
+		v.insert((asset, who), amount);
+	});
 }
 
 pub(crate) fn clear_frozen_balance(asset: u32, who: u64) {
-	FROZEN.with(|f| f.borrow_mut().remove(&(asset, who)));
+	Frozen::mutate(|v| {
+		v.remove(&(asset, who));
+	});
 }
 
 pub(crate) fn hooks() -> Vec<Hook> {
-	HOOKS.with(|h| h.borrow().clone())
+	Hooks::get().clone()
 }
 
 pub(crate) fn take_hooks() -> Vec<Hook> {
-	HOOKS.with(|h| h.take())
+	let result = Hooks::get();
+	Hooks::set(Default::default());
+	result
```

Comment:
no take api? probably we could add one

Reply:
Ok, might tackle this.

Reply:
Should be very trivial to make, similar to how I've asked you to add `::mutate`.

Reply:
I think I can implement it directly actually. Made a minor adjustment to mutate so it returns the object after it's been mutated (similar to the `.with` method)

Reply:
Doesn't work, but the method would have to be implemented on Vec, not in the macro, I believe.